[FSDP] Implement local_state_dict and load_local_state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72469

1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

fegin committed:
[FSDP] Implement local_state_dict and load_local_state_dict

1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+                    module.state_dict_type = prev_state_dict_type
+
+    @staticmethod
+    def _post_state_dict_hook(
```
Should we use some sort of basic dispatch mechanism to deconvolute / separate out implementation details from this code? 

Maybe something like 3 functions:

`def _local_post_state_dict_hook`
`def _full_post_state_dict_hook`
`def _sharded_state_dict_hook`

And in `_post_state_dict_hook`, we have:
```
dispatch: Dict[StateDictType, Callable] = { 
    StateDictType.LOCAL: _local_post_state_dict_hook,
    # and so on
}
```

and the implementation is just `return dispatch[state_dict_type](args)`

eventually, those 3 functions might be able to be moved into some sort of utils folder.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            flat_param = getattr(self.module, "flat_param", None)
+            full_numel = flat_param.full_numel
```
do we need to have `flat_param is not None` assert before this line instead of on 730?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            local_shards = [
+                Shard.from_tensor_and_offsets(flat_param, [shard_offset], self.rank)
+            ]
+            state_dict[f"{prefix}flat_param"] = init_from_local_shards(
```
nice!

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            flat_param = getattr(self.module, "flat_param", None)
```
nit: add some docs/description describing what the hook does?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            state_dict = self.module.flat_state_dict(destination, prefix, keep_vars)
+            state_dict = self._post_state_dict_hook(self, state_dict, prefix=prefix)
```
Why do we need to manually call this if it is registered as a hook in FSDP ctor?

fegin replied:
Fixed the inconsistency.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        has been wrapped with FSDP.
+        """
+        with self.set_state_dict_type(StateDictType.LOCAL_STATE_DICT):
+            return FullyShardedDataParallel.state_dict(self, *args, **kwargs)
```
will `super().state_dict()` work and automatically call all the hooks?

fegin replied:
`self.state_dict()` makes more sense. I changed it.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        will be done.
+        """
+        self = cast(FlatParameter, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
```
same comment around the dispatch idea as for `_post_state_dict_hook`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            assert len(
+                shards
+            ), "load_local_state_dict assume one shard per ShardedTensor."
+            load_tensor = cast(torch.Tensor, shards[0].tensor)
```
Can we separate out any of this logic into helper functions? Such as the process to reconstruct the sharded parameter from `shards` object. 

Also, same comment around adding comments/documentation around what this code is doing for better maintainability / understanding.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            return super().load_state_dict(state_dict, strict)
```
we use `super()` call here but don't use it in `state_dict` implementation where we actually return `FullyShardedDataParallel.state_dict()`, that seems inconsistent and can cause issues/confusion?

fegin replied:
local_state_dict() and sharded_state_dict() should all call `self.state_dict`. And self.state_dict has to call super().state_dict(). 

local_state_dict() and sharded_state_dict() are just wrapper to set state_dict_type first then call self.state_dict. The full state_dict does not have to do so because the state_dict_type defaults to STATE_DICT.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
             return False
 
+    @contextlib.contextmanager
+    def set_state_dict_type(self, state_dict_type: StateDictType) -> Generator:
```
seems only used in FSDP code and users will call things like `local_state_dict`, should we prefix the impl with `_`?

fegin replied:
This should be exported. Uses may not call `local_state_dict` directly. In fact, users cannot call `local_state_dict` directly if the FSDP module is wrapped inside another non-FSDP module, e.g., Pyper models. This allow users to do something like:
with state_dict_type(LOCAL_STATE_DICT):
       pyper_model.state_dict()

rohan-varma replied:
great point!

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -0,0 +1,127 @@
+OUTER_SHAPE = [4, 5]
+
+
+class Model(Module):
```
Wonder if we should test a more complicated set of models. A few ideas:

1. Nested FSDP module
2. FSDP module where shards need padding, i.e. total # of params is not divisible by the world size and we pad the shards. Seems like there's some specific logic there that we'd like to verify.
3. More complex module such as those with shared params, Transformer module, etc. These can be added as part of FSDP vs DDP parity test as I mentioned in another comment.

rohan-varma replied:
https://github.com/pytorch/pytorch/issues/73048

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        self = cast(FlatParameter, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
```
Logic for both local_state_dict loading and saving hooks seem nontrivial and I think they're probably worth testing as standalone functionality, which would save a bunch of time debugging if something in local_state_dict breaks down the line. What do you think?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1069,6 +1276,7 @@ def _rebuild_full_params(self) -> None:
         """
         Gather all shards of params.
         """
+        self._lazy_init()
```
Seems unrelated, why do we need to add `_lazy_init` in _rebuild_full_params now?

fegin replied:
_rebuild_full_params are used in the several our tests. In the tests, there are no guarantee that forward will be called. Adding the lazy_init should make the code more robust.

rohan-varma reviewed (commented):
This is amazing work! Please see comments inline.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives may be used.
+        """
+        torch.cuda.synchronize()
```
if torch.cuda.is_available():
            torch.cuda.synchronize()

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -0,0 +1,127 @@
+    @parametrize("state_dict_name", ["local_state_dict"])
+    def test_fsdp_state_dict_keys(self, state_dict_name):
+        state_dict = self._state_dict(self._initialize_model(True), state_dict_name)
+        if state_dict_name != "local_state_dict":
```
I remove the test for full_state_dict and sharded_state_dict. We should add the test when landing either full_state_dict or sharded_state_dict.

rohan-varma replied:
https://github.com/pytorch/pytorch/issues/73044

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -0,0 +1,127 @@
+instantiate_parametrized_tests(TestFSDPStateDict)
+
+if __name__ == "__main__":
+    run_tests()
```
Add a TODO.

fegin commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -14,20 +14,52 @@
+    module: nn.Module, state_dict: "OrderedDict[str, Tensor]", prefix: str, *args: Any
+) -> "OrderedDict[str, Tensor]":
+    # Move everything from .fpw_module up one level.
+    _replace_by_prefix(state_dict, prefix + "_fpw_module.", prefix)
```
@zhaojuanmao FlattenParamsWrapper._fpw_module can be another FSDP. If we do not apply the _replace_by_prefix, we may get keys like: {"flat_param", "_fpw_module.inner.flat_param"}. The former key is the  flat_param of the outer/parent module and the later key is the flat_param of the inner/child module.

fegin commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -387,6 +427,12 @@ def __getitem__(self, key: int) -> Any:
+        """Return the flattened state_dict."""
+        assert getattr(self, "flat_param", None) is not None
+        assert isinstance(self.flat_param, FlatParameter)
+        return self.state_dict(*args, **kwargs)
```
This is flat_state_dict() that is called when parameters are flattened. We do not override `state_dict()`, so self.state_dict(*args, **kwargs) is equivalent to  super().state_dict().

rohan-varma replied:
Is it worth to call `super().state_dict()` for consistency and avoid this confusion?

fegin replied:
I removed this API since I changed the implementation.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        A context manager to set the state_dict_type of this FSDP module and
+        its descendant FSDP modules.
+        .. note:: This API should be called for only the root FSDP module.
+        .. note:: All the FSDP modules should have the same state_dict_type.
```
Yes. Now that I added the check of is_root property, I remove the comment but still keep the check inside the API.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            local_shards = [
+                Shard.from_tensor_and_offsets(flat_param, [shard_offset], self.rank)
+            ]
+            state_dict[f"{prefix}flat_param"] = init_from_local_shards(
```
Correct, as long as `from_tensor_and_offsets()` does not copy, which I believe it does not.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            _replace_by_prefix(state_dict, prefix, f"{prefix}_fsdp_wrapped_module.")
```
Fixed the inconsistency.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+            local_shards = [
+                Shard.from_tensor_and_offsets(flat_param, [shard_offset], self.rank)
+            ]
+            state_dict[f"{prefix}flat_param"] = init_from_local_shards(
```
No, the key will actually exist and we just replace the value.  I added comment to explain the logic.

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
Looks great! Couple of minor comments and wondering if we should also test `with state_dict_type` API in addition to direct call to `local_state_dict`.

Will stamp it after CI/lint fixes

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -640,6 +696,226 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+                f"Local shard size = {flat_param.numel()} and the tensor in "
+                f"the state_dict is {load_tensor.numel()}."
+            )
+            load_tensor = F.pad(load_tensor, [0, flat_param.num_padded])
```
Not, we can abstract this in a follow up PR.

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM assuming the context manager test is added, thanks for working on this! Please ensure CI is clear before landing

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement local_state_dict and load_local_state_dict"


1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

[ghstack-poisoned]

suo commented:
Looks like this PR broke master: https://www.torch-ci.com/pytorch/pytorch/commit/d50643adcd8ca144c9b2d5de339b95be5339bf63

Unfortunately I have to revert

fegin commented:
Interesting, this PR does not touch any files listed in the tests. So it should be impossible that this PR cause those flake8 and linter errors. @suo Just curious, did reverting the PR actually recover the master?

suo commented:
the linter errors were caused by a different commit, which I also had to
revert. I believe this commit broke
https://github.com/pytorch/pytorch/runs/5260788165?check_suite_focus=true;
reverting did indeed fix master for that test

On Tue, Feb 22, 2022 at 7:45 AM Chien-Chin Huang ***@***.***>
wrote:

> Interesting, this PR does not touch any files listed in the tests. So it
> should be impossible that this PR cause those flake8 and linter errors.
> @suo <https://github.com/suo> Just curious, did reverting the PR actually
> recover the master?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/72469#issuecomment-1047475412>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMK4EAPQTURNMLVO3SYALLU4MWJ5ANCNFSM5NYV3VYQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
